File: src/chapters/iterative-methods/content/code/jacobi.r
# Solve A x = b by Jacobi iteration. Returns list(x, iterations).
jacobi <- function(A, b, x0 = NULL, tol = 1e-10, max_iter = 200) {
  A <- as.matrix(A)
  b <- as.numeric(b)
  n <- length(b)
  x <- if (is.null(x0)) rep(0, n) else as.numeric(x0)
  D <- diag(A)                 # diagonal entries
  R <- A - diag(D, n, n)       # off-diagonal part
  for (k in 1:max_iter) {
    x_new <- (b - R %*% x) / D
    x_new <- as.numeric(x_new)
    if (max(abs(x_new - x)) <= tol) {
      return(list(x = x_new, iterations = k))
    }
    x <- x_new
  }
  list(x = x, iterations = max_iter)
}

if (sys.nframe() == 0) {
  A <- rbind(c(4, 2, -1), c(5, -10, 2), c(-2, 3, -7))
  b <- c(9, 8, 3)
  res <- jacobi(A, b)
  cat("x =", res$x, " iterations =", res$iterations, "\n")
}

File: src/chapters/iterative-methods/content/code/jacobi.m
function [x, k] = jacobi(A, b, x0, tol, max_iter)
% JACOBI  Solve A x = b by Jacobi iteration.  Returns x and iteration count k.
    if nargin < 3, x0 = zeros(size(b)); end
    if nargin < 4, tol = 1e-10; end
    if nargin < 5, max_iter = 200; end

    D = diag(diag(A));            % diagonal part
    T = eye(numel(b)) - D \ A;    % iteration matrix  T = I - D^{-1} A
    c = D \ b;                    % preconditioned right-hand side

    x = x0;
    for k = 1:max_iter
        x_new = T * x + c;        % x^{(k+1)} = T x^{(k)} + c
        if norm(x_new - x, inf) <= tol
            x = x_new; return
        end
        x = x_new;
    end
end

% --- Demo ---
A = [4 2 -1; 5 -10 2; -2 3 -7];
b = [9; 8; 3];
[x, k] = jacobi(A, b);
fprintf('x = [%g %g %g],  iterations = %d\n', x(1), x(2), x(3), k);
